man/mu-find.1
.TH MU FIND 1 "29 April 2022" "User Manuals"

.SH NAME

mu find \- find e-mail messages in the \fBmu\fR database.

.SH SYNOPSIS

.B mu find [options] <search expression>

.SH DESCRIPTION

\fBmu find\fR is the \fBmu\fR command for searching e-mail message
that were stored earlier using \fBmu index\fR(1).

.SH SEARCHING MAIL

\fBmu find\fR starts a search for messages in the database that match
some search pattern. The search patterns are described in detail in
.BR mu-query (7).
.

For example:

.nf
   $ mu find subject:snow and date:2009..
.fi

would find all messages in 2009 with 'snow' in the subject field, e.g:

.nf
  2009-03-05 17:57:33 EET Lucia  <lucia@example.com> running in the snow
  2009-03-05 18:38:24 EET Marius <marius@foobar.com> Re: running in the snow
.fi

Note, this the default, plain-text output, which is the default, so you don't
have to use \fB--format=plain\fR. For other types of output (such as symlinks,
XML or s-expressions), see the discussion in the \fBOPTIONS\fR-section below
about \fB--format\fR.

The search pattern is taken as a command-line parameter. If the search
parameter consists of multiple parts (as in the example) they are
treated as if there were a logical \fBand\fR between them.

For details on the possible queries, see
.BR mu-query (7).

.SH OPTIONS

Note, some of the important options are described in the \fBmu\fR(1) man-page
and not here, as they apply to multiple mu-commands.

The \fBfind\fR-command has various options that influence the way \fBmu\fR
displays the results. If you don't specify anything, the defaults are
\fI\-\-fields="d f s"\fR, \fI\-\-sortfield=date\fR and \fI\-\-reverse\fR.

.TP
\fB\-f\fR, \fB\-\-fields\fR=\fI<fields>\fR
specifies a string that determines which fields are shown in the output. This
string consists of a number of characters (such as 's' for subject or 'f' for
from), which will replace with the actual field in the output. Fields that are
not known will be output as-is, allowing for some simple formatting.

For example:

.nf
  $ mu find subject:snow --fields "d f s"
.fi

would list the date, subject and sender of all messages with 'snow' in the
their subject.

The table of replacement characters is superset of the list mentions for
search parameters, such as:
.nf
	t       \fBt\fRo: recipient
	d       Sent \fBd\fRate of the message
	f       Message sender (\fBf\fRrom:)
	g       Message flags (fla\fBg\fRs)
	l       Full path to the message (\fBl\fRocation)
	s       Message \fBs\fRubject
	i       Message-\fBi\fRd
	m       \fBm\fRaildir
.fi

For the complete, up-to-date list, see:
.BR mu-fields(1)

The message flags are described in \fBmu-query\fR(7). As an example, a
message which is 'seen', has an attachment and is signed would
have 'asz' as its corresponding output string, while an encrypted new
message would have 'nx'.

.TP
\fB\-s\fR, \fB\-\-sortfield\fR \fR=\fI<field>\fR and \fB\-z\fR,
\fB\-\-reverse\fR specifies the field to sort the search results by, and the
direction (i.e., 'reverse' means that the sort should be reverted - Z-A). Examples include:

.nf
	cc,c            Cc (carbon-copy) recipient(s)
	date,d          Message sent date
	from,f          Message sender
	maildir,m       Maildir
	msgid,i         Message id
	prio,p          Nessage priority
	subject,s       Message subject
	to,t            To:-recipient(s)
.fi

For the complete list use can use the \fBmu fields\fR command; see:
.BR mu-fields(1)

Thus, for example, to sort messages by date, you could specify:

.nf
  $ mu find fahrrad --fields "d f s" --sortfield=date --reverse
.fi

Note, if you specify a sortfield, by default, messages are sorted in reverse
(descending) order (e.g., from lowest to highest). This is usually a good
choice, but for dates it may be more useful to sort in the opposite direction.

.TP
\fB\-n\fR, \fB\-\-maxnum=<number>\fR
If > 0, display maximally that number of entries.  If not specified, all matching entries are displayed.

.TP
\fB\-\-summary-len=<number>\fR
If > 0, use that number of lines of the message to provide a summary.

.TP
\fB\-\-format\fR=\fIplain|links|xquery|xml|sexp\fR
output results in the specified format.

The default is \fBplain\fR, i.e normal output with one line per message.

\fBlinks\fR outputs the results as a maildir with symbolic links to the found
messages. This enables easy integration with mail-clients (see below for more
information).

\fBxml\fR formats the search results as XML.

\fBsexp\fR formats the search results as an s-expression as used in Lisp
programming environments.

\fBxquery\fR shows the Xapian query corresponding to your search terms. This
is meant for for debugging purposes.

.TP
\fB\-\-linksdir\fR \fR=\fI<dir>\fR and \fB\-c\fR, \fB\-\-clearlinks\fR
output the results as a maildir with symbolic links to the found
messages. This enables easy integration with mail-clients (see below
for more information). \fBmu\fR will create the maildir if it does not
exist yet.

If you specify \fB\-\-clearlinks\fR, all existing symlinks will be
cleared from the target directories; this allows for re-use of the
same maildir. However, this option will delete any symlink it finds,
so be careful.

.nf
  $ mu find grolsch --format=links --linksdir=~/Maildir/search --clearlinks
.fi

will store links to found messages in \fI~/Maildir/search\fR. If the directory
does not exist yet, it will be created.

Note: when \fBmu\fR creates a Maildir for these links, it automatically
inserts a \fI.noindex\fR file, to exclude the directory from \fBmu
index\fR.

.TP
\fB\-\-after=\fR\fI<timestamp>\fR only show messages whose message files were
last modified (\fBmtime\fR) after \fI<timestamp>\fR. \fI<timestamp>\fR is a
UNIX \fBtime_t\fR value, the number of seconds since 1970-01-01 (in UTC).

From the command line, you can use the \fBdate\fR command to get this
value. For example, only consider messages modified (or created) in the last 5
minutes, you could specify
.nf
  --after=`date +%s --date='5 min ago'`
.fi
This is assuming the GNU \fBdate\fR command.


.TP
\fB\-\-exec\fR=\fI<command>\fR
the \fB\-\-exec\fR command causes the \fIcommand\fR to be executed on each
matched message; for example, to see the raw text of all messages
matching 'milkshake', you could use:
.nf
  $ mu find milkshake --exec='less'
.fi
which is roughly equivalent to:
.nf
  $ mu find milkshake --fields="l" | xargs less
.fi


.TP
\fB\-b\fR, \fB\-\-bookmark\fR=\fI<bookmark>\fR
use a bookmarked search query. Using this option, a query from your bookmark
file will be prepended to other search queries. See \fBmu-bookmarks\fR(1) for the
details of the bookmarks file.


.TP
\fB\-\-skip\-dups\fR,\fB-u\fR whenever there are multiple messages with the
same name, only show the first one. This is useful if you have copies of the
same message, which is a common occurrence when using e.g. Gmail together with
\fBofflineimap\fR.

.TP
\fB\-\-include\-related\fR,\fB-r\fR also include messages being referred to by
the matched messages -- i.e.. include messages that are part of the same
message thread as some matched messages. This is useful if you want
Gmail-style 'conversations'. Note, finding these related messages make
searches slower.

.TP
\fB\-t\fR, \fB\-\-threads\fR show messages in a 'threaded' format --
that is, with indentation and arrows showing the conversation threads
in the list of matching messages. When using this, sorting is
chronological (by date), based on the newest message in a thread.

Messages in the threaded list are indented based on the depth in the
discussion, and are prefix with a kind of arrow with thread-related
information about the message, as in the following table:

.nf
|             | normal | orphan | duplicate |
|-------------+--------+--------+-----------|
| first child | `->    | `*>    | `=>       |
| other       | |->    | |*>    | |=>       |
.fi

Here, an 'orphan' is a message without a parent message (in the list of
matches), and a duplicate is a message whose message-id was already seen
before; not this may not really be the same message, if the message-id was
copied.

The algorithm used for determining the threads is based on Jamie Zawinksi's
description:
.BR http://www.jwz.org/doc/threading.html


.SS Integrating mu find with mail clients

.TP

\fBmutt\fR

For \fBmutt\fR you can use the following in your \fImuttrc\fR; pressing the F8
key will start a search, and F9 will take you to the results.

.nf
# mutt macros for mu
macro index <F8> "<shell-escape>mu find --clearlinks --format=links --linksdir=~/Maildir/search " \\
					"mu find"
macro index <F9> "<change-folder-readonly>~/Maildir/search" \\
					"mu find results"
.fi


.TP

\fBWanderlust\fR

\fBSam B\fR suggested the following on the \fBmu\fR-mailing list. First add
the following to your Wanderlust configuration file:

.nf
(require 'elmo-search)
(elmo-search-register-engine
    'mu 'local-file
    :prog "/usr/local/bin/mu" ;; or wherever you've installed it
    :args '("find" pattern "--fields" "l") :charset 'utf-8)

(setq elmo-search-default-engine 'mu)
;; for when you type "g" in folder or summary.
(setq wl-default-spec "[")
.fi

Now, you can search using the \fBg\fR key binding; you can also create
permanent virtual folders when the messages matching some expression by adding
something like the following to your \fIfolders\fR file.

.nf
VFolders {
  [date:today..now]!mu  "Today"

  [size:1m..100m]!mu    "Big"

  [flag:unread]!mu      "Unread"
}
.fi

After restarting Wanderlust, the virtual folders should appear.

.SH RETURN VALUE

\fBmu find\fR returns 0 upon successful completion; if the search was
performed, there needs to be a least one match. Anything else leads to a
non-zero return value, for example:

.nf
| code | meaning                        |
|------+--------------------------------|
|    0 | ok                             |
|    1 | general error                  |
|    4 | no matches (for 'mu find')     |
.fi


.SH ENCODING

\fBmu find\fR output is encoded according the locale for \fI--format=plain\fR
(the default), and UTF-8 for all other formats (\fIsexp\fR,
\fIxml\fR).


.SH BUGS

Please report bugs if you find them:
.BR https://github.com/djcb/mu/issues
If you have specific messages which are not matched correctly, please attach
them (appropriately censored if needed).

.SH AUTHOR

Dirk-Jan C. Binnema <djcb@djcbsoftware.nl>

.SH "SEE ALSO"

.BR mu (1),
.BR mu-index (1),
.BR mu-query (7)
.BR mu-fields (1)


NEWS
See NEWS.org



emacs/mu-find.el
;;; mu-find.el -- use `mu' from emacs
;; Copyright (C) 2011 Dirk-Jan C. Binnema

;; Author: Dirk-Jan C. Binnema <djcb@djcbsoftware.nl>
;; Maintainer: Dirk-Jan C. Binnema <djcb@djcbsoftware.nl>
;; Keywords: email
;; Version: 0.0

;; This file is not part of GNU Emacs.
;;
;; GNU Emacs is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; GNU Emacs is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;;; Code:

(require 'mu-common)

(defvar mu-find-fields
  '( (:date          .  25)
     (:from-or-to    .  22)
     (:subject       .  40))
  "a list of fields and their widths")

(defvar mu-find-sort-field "date"
  "shortcut of the field to sort on (see mu-find (1))")
(defvar mu-find-sort-descending nil
  "whether to sort in descending order")

;; internal stuff
(defconst mu-find-buffer-name " *mu-find*" "name of the mu
results buffer; name should start with a space")
(defvar mu-find-process nil "the possibly running find process")
(defconst mu-find-process-name "*<mu-find-process>*" "name of the mu
results buffer; name should start with a space")
(defconst mu-eom "\n;;eom\n" "marker for the end of message in
the mu find output")
(defvar mu-find-expression nil
  "search expression for the current find buffer")

(defvar mu-buf "" "buffer for results data")
(defun mu-find-process-filter (proc str)
  "process-filter for the 'mu find --format=sexp output; it
  accumulates the strings into valid sexps by checking of the
  ';;eom' end-of-msg marker, and then evaluating them"
  (with-current-buffer mu-find-buffer-name
    (save-excursion
      (setq mu-buf (concat mu-buf str))
      (let ((eom (string-match mu-eom mu-buf)))
	(while (numberp eom)
	  (let* ((msg (car (read-from-string (substring mu-buf 0 eom))))
		  (inhibit-read-only t))
	    (goto-char (point-max))
	    (save-match-data (insert (mu-find-header msg) ?\n)))
	  (setq mu-buf (substring mu-buf (match-end 0)))
	  (setq eom (string-match mu-eom mu-buf)))))))


(defun mu-find-process-sentinel (proc msg)
  "Check the mu-find process upon completion"
  (let ((status (process-status proc))
	 (exit-status (process-exit-status proc)))
    (if (memq status '(exit signal))
      (let ((inhibit-read-only t)
	     (text
	       (cond
		 ((eq status 'signal)
		   "Search process killed (results incomplete)")
		 ((eq status 'exit)
		   (cond
		     ((= 0 exit-status) "End of search results")
		     ((= 2 exit-status) "No matches found")
		     ((= 4 exit-status) "Database problem; try running 'mu index'")
		     (t (format "Some error occured; mu-find returned %d"
			  exit-status))))
		 (t "Unknown status")))) ;; shouldn't happen
	(when (get-buffer mu-find-buffer-name)
		(with-current-buffer mu-find-buffer-name
		  (save-excursion
		    (goto-char (point-max))
		    (insert (mu-str text)))))))))


;; Note, the 'mu find --format=sexp' sexp is almost the same as the ones that
;; 'mu view --format=sexp' produces (see mu-get-message), with the difference
;; that former may give more than one result, and that mu-find output comes from
;; the database rather than file, and does _not_ contain the message body
(defun mu-find (expr)
  "search in the mu database"
  (interactive "s[mu] match expr: ")
  (let* ((output (get-buffer mu-find-buffer-name)))
    (when output (kill-buffer output))
    (setq output (get-buffer-create mu-find-buffer-name) mu-buf "")
    (let* ((dummy-arg "--fields=\"dummy\"") ;; ignored
	    (proc
	      (start-process mu-find-process-name mu-find-process-name
		mu-binary
		"find"
		(if mu-home
		  (concat "--muhome=" mu-home) dummy-arg)
		(if mu-find-sort-field
		  (concat "--sortfield=" mu-find-sort-field) dummy-arg)
		(if mu-find-sort-descending "--descending" dummy-arg)
		"--format=sexp"
		"--quiet"
	      expr)))
      (set-process-filter   proc 'mu-find-process-filter)
      (set-process-sentinel proc 'mu-find-process-sentinel)
      (setq mu-find-process proc)
      (switch-to-buffer output)
      (setq mu-find-expression expr)
      ;; (make-variable-buffer-local mu-find-expression)
      (mu-find-mode))))

(defun mu-find-display-contact (lst width face)
  "display a list of contacts, truncated for fitting in WIDTH"
  (if lst
    (let* ((len (length lst))
	    (str (if (= len 0) "<none>"
		   ;; try name -> email -> ?
		   (or (car(car lst)) (cdr(car lst)) "?")))		   
	    (others (if (> len 1) (mu-str (format " [+%d]" (- len 1))) "")))
      (truncate-string-to-width 
	(concat(propertize (truncate-string-to-width str
			     (- width (length others)) 0 ?\s "...") 'face face) others)
	width 0 ?\s))
    (make-string width ?\s)))


(defun mu-find-display-from-or-to (fromlst tolst width from-face to-face)
  "return a propertized string for FROM unless TO matches
  mu-own-address, in which case it returns TO, prefixed with To:"
  (if (and fromlst tolst)
    (let ((fromaddr (cdr(car fromlst))))
      (if (and fromaddr (string-match mu-own-address fromaddr))
	(concat (mu-str "To ") (mu-find-display-contact tolst (- width 3) to-face))
	(mu-find-display-contact fromlst width from-face)))
    (make-string width ?\s)))

(defun mu-find-display-size (size width face)
  "return a string for SIZE of WIDTH with FACE"
  (let* ((str
	   (cond
	     ((>= size 1000000) (format "%2.1fM" (/ size 1000000.0)))
	     ((and (>= size 1000) (< size 1000000)) (format "%2.1fK" (/ size 1000.0)))
	     ((< size 1000) (format "%d" size)))))
    (propertize  (truncate-string-to-width str width 0 ?\s) 'face face)))
	  

(defun mu-find-display-str (str width face)
  "print a STR, at WIDTH (truncate or ' '-pad) with FACE"
  (let ((str (if str str "")))
    (propertize (truncate-string-to-width str width 0 ?\s t) 'face face)))

(defun mu-find-display-flags (flags width face)
  (let ((str
	  (mapconcat
	    (lambda(flag)
	      (let ((flagname (symbol-name flag)))
		(cond 
		  ((string= flagname "unread")    "U")
		  ((string= flagname "seen")      "S")
		  ((string= flagname "replied")   "R")
		  ((string= flagname "attach")    "a")
		  ((string= flagname "encrypted") "x")
		  ((string= flagname "signed")    "s")))) flags "")))
    (propertize  (truncate-string-to-width str width 0 ?\s) 'face face)))

(defun mu-find-header (msg)
  "convert a message s-expression into a header for display"
  (let ((hdr (concat "  " (mapconcat
	 (lambda (fieldinfo)
	   (let ((field (car fieldinfo)) (width (cdr fieldinfo)))
	     (case field  
	       (:date
		 (mu-find-display-str (format-time-string mu-date-format-short
					(plist-get msg :date)) width 'mu-date-face))	  
	       (:from
		 (mu-find-display-contact (plist-get msg :from) width 'mu-from-face))
	       (:to
		 (mu-find-display-contact (plist-get msg :to) width 'mu-to-face))
	       (:cc
		 (mu-find-display-contact (plist-get msg :cc) width 'mu-cc-face))	    
	       (:bcc
		 (mu-find-display-contact (plist-get msg :bcc) width 'mu-bcc-face))
	       (:flags
		 (mu-find-display-flags (plist-get msg :flags) width 'mu-flag-face))
	       (:size
		 (mu-find-display-size (plist-get msg :size) width 'mu-size-face))
	       (:from-or-to
		 (mu-find-display-from-or-to (plist-get msg :from)
		   (plist-get msg :to) width 'mu-from-face 'mu-to-face))
	       (:subject
		 (mu-find-display-str (plist-get msg :subject) width
		   'mu-subject-face)))))
	 mu-find-fields " "))))
    (setq hdr (mu-find-set-props-for-flags hdr (plist-get msg :flags)))
    (propertize hdr 'path (plist-get msg :path) 'front-sticky t)))

(defun mu-find-set-props-for-flags (hdr flags)
  "set text properties/faces based on flags"
  (if (memq 'unread flags)
    (add-text-properties 0 (- (length hdr) 1) '(face (:weight bold)) hdr))
    hdr)


(defun mu-find-mode ()
  "major mode for displaying search results"
  (interactive)
  (kill-all-local-variables)
  (use-local-map mu-find-mode-map)
  (setq
    major-mode 'mu-find-mode mode-name "*headers*"
    truncate-lines t buffer-read-only t
    overwrite-mode 'overwrite-mode-binary))

(defvar mu-find-mode-map
  (let ((map (make-sparse-keymap)))
    (define-key map "q" 'mu-quit-buffer)
    (define-key map "s" 'mu-find-change-sort)
    (define-key map "g" 'mu-find-refresh)

    ;; marking/unmarking
    (define-key map "m" 'mu-find-mark-for-move)
    (define-key map "d" 'mu-find-mark-for-trash)
    (define-key map "D" 'mu-find-mark-for-deletion)
    (define-key map "u" 'mu-find-unmark)

    ;; message composition
    (define-key map "r" 'mu-reply)
    (define-key map "f" 'mu-forward)
    (define-key map (kbd "RET") 'mu-find-view)
    map)
  "Keymap for \"mu-find\" buffers.")
(fset 'mu-find-mode-map mu-find-mode-map)

(defun mu-find-view ()
  "display the message at the current line"
  (interactive)
  (let ((path (mu-get-path)))
    (when path (mu-view path))))

(defun mu-find-next  ()
  "go to the next line; t if it worked, nil otherwise"
  (interactive)
  (if (or (/= 0 (forward-line 1)) (not (mu-get-path)))    
    (progn (message "No message after this one") nil)
    t))

(defun mu-find-prev  ()
  "go to the next line; t if it worked, nil otherwise"
  (interactive)
  (if (/= 0 (forward-line -1))
    (progn (message "No message before this one") nil)
    t))

(defun mu-find-refresh ()
  "re-run the query for the current search expression"
  (interactive) 
  (unless (and mu-find-process
	    (eq (process-status mu-find-process) 'run))
    (when mu-find-expression
      (mu-find mu-find-expression))))

(defun mu-find-change-sort-order (fieldchar)
  "change the sortfield to FIELDCHAR"
  (interactive"cField to sort by ('d', 's', etc.; see mu-find(1)):\n")
  (let
    ((field
       (case fieldchar
	 (?b "bcc")
	 (?c "cc")
	 (?d "date")
	 (?f "from")
	 (?i "msgid")
	 (?m "maildir")
	 (?p "prio")
	 (?s "subject")
	 (?t "to")
	 (?z "size"))))
    (if field
      (setq mu-find-sort-field field)
      (message "Invalid sort-field; use one of bcdfimpstz (see mu-find(1)"))
    field))

(defun mu-find-change-sort-direction (dirchar)
  "change the sort direction, either [a]scending or [d]escending"
  (interactive
    "cSorting direction ([a]scending or [d]escending):")
  (cond
    (?d (setq mu-find-sort-descending t) t)
    (?a (setq mu-find-sort-descending nil) t)
    (t 	(message
	  "Invalid sort-direction; choose either [a]scending or [d]escending") nil)))

(defun mu-find-mark (what)
 "mark the current msg for 'trash, 'move, 'none"
  (when (mu-get-path)
    (move-beginning-of-line 1)
    (let ((inhibit-read-only t) (overwrite-mode nil))
      (if (and (not (eq what 'none)) (get-text-property (point) 'action))
	(message "Message at pooint is already marked")
	(progn
	  (delete-char 1)
	  (case what
	    ('trash (insert-and-inherit
		      (mu-str (propertize "d" 'action what 'target "/foo/bar"))))
	    ('delete (insert-and-inherit
		      (mu-str (propertize "D" 'action what 'target "/foo/bar"))))
	    ('move  (insert-and-inherit
		      (mu-str (propertize "m" 'action what 'target "/foo/bar"))))
	    ('none  (insert-and-inherit " ")))
	  (forward-line))))))
  
(defun mu-find-mark-for-trash ()
  (interactive)
  (mu-find-mark 'trash))

(defun mu-find-mark-for-deletion ()
  (interactive)
  (mu-find-mark 'delete))

(defun mu-find-mark-for-move ()
  (interactive)
  (mu-find-mark 'move))

(defun mu-find-unmark ()
  (interactive)
  (mu-find-mark 'none))

(defun mu-find-change-sort ()
  "change sort field and direction"
  (interactive)
  (and (call-interactively 'mu-find-change-sort-order)
    (call-interactively 'mu-find-change-sort-direction)))

(provide 'mu-find)
  
